pkg/commands/credentials.go

```diff
@@ -0,0 +1,366 @@
+		if err != nil {
+			end <- err
+			return
+		}
+
+		serverStartedChan <- struct{}{}
+		rpc.Accept(inbound)
+
+		serverRunning = false
+	}()
+
+	err = <-end
+	if serverRunning {
+		inbound.Close()
+	}
+
+	return err
+}
+
+// GetFreePort returns a free port that can be used by lazygit
+func GetFreePort() string {
+	checkFrom := 5000
+	toReturn := ""
+	for {
+		checkFrom++
+		check := fmt.Sprintf("%v", checkFrom)
+		if IsFreePort(check) {
+			toReturn = check
+			break
+		}
+	}
+	return toReturn
+}
+
+// IsFreePort return true if the port if not in use
+func IsFreePort(port string) bool {
+	conn, err := net.Dial("tcp", "127.0.0.1:"+port)
+	if err == nil {
+		go conn.Close()
+		return false
+	}
+	return true
+}
+
+// SetupClient sets up the client
+// This will be called if lazygit is called through git
+func SetupClient() {
+	hostPubText := os.Getenv("LAZYGIT_HOST_PUBLIC_KEY")
+	hostPort := os.Getenv("LAZYGIT_HOST_PORT")
+	listenerNumber := os.Getenv("LAZYGIT_LISTENER")
+
+	hostPubBlock, _ := pem.Decode([]byte(hostPubText))
+	hostPub, err := x509.ParsePKCS1PublicKey(hostPubBlock.Bytes)
+	if err != nil {
+		return
+	}
+
+	clientPriv, err := rsa.GenerateKey(rand.Reader, 3072)
+	if err != nil {
+		return
+	}
+	clientPub := clientPriv.PublicKey
+	clientPubText := pem.EncodeToMemory(&pem.Block{Type: "RSA PUBLIC KEY", Bytes: x509.MarshalPKCS1PublicKey(&clientPub)})
+	toSend := fmt.Sprintf("%x|%v", sha256.Sum256(clientPubText), os.Args[1])
+	encryptedData, err := rsa.EncryptOAEP(sha256.New(), rand.Reader, hostPub, []byte(toSend), []byte("TELL HOST"))
+	if err != nil {
+		return
+	}
+
+	rply, err := SendToLG(hostPort, listenerNumber, "Input", InputQuestion{
+		ClientPublicKeyText: string(clientPubText),
+		Message:             encryptedData,
+		ListenerKey:         listenerNumber,
+	})
+	if err != nil {
+		return
+	}
+
+	msg, err := rsa.DecryptOAEP(sha256.New(), rand.Reader, clientPriv, rply, []byte("TO PRINT"))
+	if err != nil {
+		return
+	}
+
+	fmt.Println(string(msg))
+}
+
+// SendToLG sends a message to the lazygit host
+func SendToLG(port, listenerNumber string, selectFunction string, args interface{}) ([]byte, error) {
+	client, err := rpc.Dial("tcp", "127.0.0.1:"+port)
+	if err != nil {
+		return nil, err
+	}
+	var out *[]byte
+	err = client.Call("Listener"+listenerNumber+"."+selectFunction, args, &out)
+	client.Close()
+	return *out, err
+}
+
+// HasLGAsSubProcess returns true if lazygit is a child of this process
+func HasLGAsSubProcess() bool {
+	lgHostPid := os.Getpid()
+	list, err := ps.Processes()
+	if err != nil {
+		log.Fatal(err)
+	}
+	status := false
+procListLoop:
+	for _, proc := range list {
+		procName := proc.Executable()
+		if procName != "lazygit" {
+			continue
+		}
+		stepsBack := 1
+		parrent := proc.PPid()
+		for {
+			proc, err := ps.FindProcess(parrent)
+			if err != nil {
+				continue procListLoop
+			}
+			if proc.Pid() == lgHostPid {
+				status = true
+				break procListLoop
+			}
+			stepsBack++
+			if stepsBack > 3 {
```

Comment:
If I understand correctly, we would expect the chain to go `lazygit -> git -> lazygit`. Is that why we go back three steps? If so, I'm not sure why we would want to accept `lazygit -> lazygit` as well.

Reply:
That's what i though to but it usually looks more like this: `lazygit -> git -> git -> lazygit`.

This check is not a security measure it's more that the package gives a `index out of range` panic when there is no PPID of the current for loop's PID.